marimo/_server/ai/providers.py

```diff
@@ -75,79 +75,101 @@ class AnyProviderConfig:
-            base_url=_get_base_url(config["anthropic"]),
+    @classmethod
+    def for_anthropic(cls, config: AiConfig) -> AnyProviderConfig:
+        ai_config = _get_ai_config(config, "anthropic", "Anthropic")
+        key = _get_key(ai_config, "Anthropic")
+        return cls(
+            base_url=_get_base_url(ai_config),
             api_key=key,
         )
 
-    @staticmethod
-    def for_google(config: AiConfig) -> AnyProviderConfig:
-        if "google" not in config:
-            raise HTTPException(
-                status_code=HTTPStatus.BAD_REQUEST,
-                detail="Google config not found",
-            )
-        key = _get_key(config["google"], "Google AI")
-        return AnyProviderConfig(
-            base_url=_get_base_url(config["google"]),
+    @classmethod
+    def for_google(cls, config: AiConfig) -> AnyProviderConfig:
+        ai_config = _get_ai_config(config, "google", "Google AI")
+        key = _get_key(ai_config, "Google AI")
+        return cls(
+            base_url=_get_base_url(ai_config),
             api_key=key,
         )
 
-    @staticmethod
-    def for_bedrock(config: AiConfig) -> AnyProviderConfig:
-        if "bedrock" not in config:
-            raise HTTPException(
-                status_code=HTTPStatus.BAD_REQUEST,
-                detail="Bedrock config not found",
-            )
-        key = _get_key(config["bedrock"], "Bedrock")
-        return AnyProviderConfig(
+    @classmethod
+    def for_bedrock(cls, config: AiConfig) -> AnyProviderConfig:
+        ai_config = _get_ai_config(config, "bedrock", "Bedrock")
+        key = _get_key(ai_config, "Bedrock")
+        return cls(
             base_url=_get_base_url(config["bedrock"], "Bedrock"),
             api_key=key,
         )
 
-    @staticmethod
-    def for_completion(config: CompletionConfig) -> AnyProviderConfig:
+    @classmethod
+    def for_completion(cls, config: CompletionConfig) -> AnyProviderConfig:
         key = _get_key(config, "AI completion")
-        return AnyProviderConfig(
+        return cls(
             base_url=_get_base_url(config),
             api_key=key,
         )
 
-    @staticmethod
-    def for_model(model: str, config: AiConfig) -> AnyProviderConfig:
+    @classmethod
+    def for_model(cls, model: str, config: AiConfig) -> AnyProviderConfig:
         if _model_is_anthropic(model):
-            return AnyProviderConfig.for_anthropic(config)
+            return cls.for_anthropic(config)
         elif _model_is_google(model):
-            return AnyProviderConfig.for_google(config)
+            return cls.for_google(config)
         elif _model_is_bedrock(model):
-            return AnyProviderConfig.for_bedrock(config)
+            return cls.for_bedrock(config)
         else:
-            return AnyProviderConfig.for_openai(config)
+            return cls.for_openai(config)
+
+    @classmethod
+    def from_provider_type(cls, config: AiConfig) -> AnyProviderConfig | None:
```

Comment:
i guess for backwards compat we can check if `provider_type` is defined and if so, use `from_provider_type` else use `for_model`?

Reply:
Yes! I've added it for the `/chat` endpoint.